File: src/core/ledger.ts
import fs from "node:fs";
import path from "node:path";

import type { RuntimeName, UsageSnapshot, UsageTotals } from "./types.js";

interface UsageLedgerRecord {
  schemaVersion: 1;
  id: string;
  runtime: RuntimeName;
  timestamp: number;
  usage: UsageSnapshot;
}

interface NormalizedUsageLedgerRecord {
  id: string;
  runtime: RuntimeName;
  timestamp: number;
  tokens: number;
  cost: number;
  currency?: "CNY" | "USD";
}

function dateKey(timestamp: number, timezone: string): string {
  const parts = new Intl.DateTimeFormat("en-CA", {
    timeZone: timezone,
    year: "numeric",
    month: "2-digit",
    day: "2-digit",
  }).formatToParts(timestamp);
  const get = (type: string): string =>
    parts.find((part) => part.type === type)?.value ?? "";
  return `${get("year")}-${get("month")}-${get("day")}`;
}

function objectRecord(value: unknown): Record<string, unknown> {
  return value && typeof value === "object" && !Array.isArray(value)
    ? (value as Record<string, unknown>)
    : {};
}

function finiteNumber(...values: unknown[]): number | undefined {
  return values.find(
    (value): value is number =>
      typeof value === "number" && Number.isFinite(value),
  );
}

function normalizeRecord(
  value: unknown,
): NormalizedUsageLedgerRecord | undefined {
  if (!value || typeof value !== "object") {
    return undefined;
  }
  const record = objectRecord(value);
  const runtime = record.runtime;
  const id = typeof record.id === "string" ? record.id : record.sessionId;
  const timestamp = finiteNumber(record.timestamp);
  if (
    typeof id !== "string" ||
    (runtime !== "openclaw" && runtime !== "hermes") ||
    timestamp === undefined
  ) {
    return undefined;
  }
  const usage = objectRecord(record.usage);
  const inputTokens = finiteNumber(usage.inputTokens, usage.input_tokens) ?? 0;
  const outputTokens =
    finiteNumber(usage.outputTokens, usage.output_tokens) ?? 0;
  const cacheReadTokens =
    finiteNumber(usage.cacheReadTokens, usage.cache_read_tokens) ?? 0;
  const cacheWriteTokens =
    finiteNumber(usage.cacheWriteTokens, usage.cache_write_tokens) ?? 0;
  const tokens =
    finiteNumber(record.tokens, usage.totalTokens, usage.total_tokens) ??
    inputTokens + outputTokens + cacheReadTokens + cacheWriteTokens;
  const cost = finiteNumber(record.cost, usage.turnCost, usage.turn_cost) ?? 0;
  const rawCurrency = record.currency ?? usage.currency;
  const currency =
    rawCurrency === "CNY" || rawCurrency === "USD" ? rawCurrency : undefined;
  return {
    id,
    runtime,
    timestamp,
    tokens: Math.max(0, tokens),
    cost: Math.max(0, cost),
    ...(currency ? { currency } : {}),
  };
}

export class UsageLedger {
  readonly filePath: string;
  private readonly timezone: string;
  private readonly records = new Map<string, NormalizedUsageLedgerRecord>();

  constructor(params: { storageDir: string; timezone: string }) {
    this.filePath = path.join(params.storageDir, "usage.ndjson");
    this.timezone = params.timezone;
    fs.mkdirSync(params.storageDir, { recursive: true, mode: 0o700 });
    this.load();
  }

  append(params: {
    id: string;
    runtime: RuntimeName;
    usage: UsageSnapshot;
    timestamp?: number;
  }): boolean {
    this.load();
    if (this.records.has(params.id)) {
      return false;
    }
    const record: UsageLedgerRecord = {
      schemaVersion: 1,
      id: params.id,
      runtime: params.runtime,
      timestamp: params.timestamp ?? Date.now(),
      usage: params.usage,
    };
    fs.appendFileSync(this.filePath, `${JSON.stringify(record)}\n`, {
      encoding: "utf8",
      mode: 0o600,
      flag: "a",
    });
    const normalized = normalizeRecord(record);
    if (normalized) {
      this.records.set(record.id, normalized);
    }
    return true;
  }

  totals(now = Date.now()): UsageTotals {
    this.load();
    const today = dateKey(now, this.timezone);
    const month = today.slice(0, 7);
    const totals: UsageTotals = {
      todayTokens: 0,
      monthTokens: 0,
      allTimeTokens: 0,
      todayCost: 0,
      monthCost: 0,
      allTimeCost: 0,
    };
    const currencies = new Set<"CNY" | "USD">();
    for (const record of this.records.values()) {
      const key = dateKey(record.timestamp, this.timezone);
      totals.allTimeTokens += record.tokens;
      totals.allTimeCost += record.cost;
      if (key.startsWith(month)) {
        totals.monthTokens += record.tokens;
        totals.monthCost += record.cost;
      }
      if (key === today) {
        totals.todayTokens += record.tokens;
        totals.todayCost += record.cost;
      }
      if (record.currency) {
        currencies.add(record.currency);
      }
    }
    const currency = currencies.size === 1 ? [...currencies][0] : undefined;
    if (currency) {
      totals.currency = currency;
    }
    return totals;
  }

  private load(): void {
    if (!fs.existsSync(this.filePath)) {
      return;
    }
    const lines = fs.readFileSync(this.filePath, "utf8").split(/\r?\n/);
    for (const line of lines) {
      if (!line.trim()) {
        continue;
      }
      try {
        const record = normalizeRecord(JSON.parse(line));
        if (record) {
          this.records.set(record.id, record);
        }
      } catch {
        // A truncated tail can occur after an abrupt process exit. Earlier
        // complete records remain authoritative.
      }
    }
  }
}

File: openclaw_hermes_feishu_card/ledger.py
from __future__ import annotations

import json
from contextlib import suppress
from dataclasses import dataclass
from datetime import UTC, datetime, tzinfo
from math import isfinite
from pathlib import Path
from threading import RLock
from zoneinfo import ZoneInfo, ZoneInfoNotFoundError

from .models import Currency, UsageSnapshot, UsageTotals


@dataclass(frozen=True, slots=True)
class _LedgerRecord:
    timestamp: float
    tokens: int
    cost: float
    currency: Currency | None


class UsageLedger:
    def __init__(self, storage_dir: Path, timezone: str) -> None:
        self._path = storage_dir / "usage.ndjson"
        resolved_timezone: tzinfo = UTC
        with suppress(ZoneInfoNotFoundError):
            resolved_timezone = ZoneInfo(timezone)
        self._timezone = resolved_timezone
        self._lock = RLock()

    def append(
        self,
        runtime: str,
        session_id: str,
        usage: UsageSnapshot,
        timestamp_ms: int,
    ) -> bool:
        record = {
            "schemaVersion": 1,
            "id": session_id,
            "runtime": runtime,
            "timestamp": timestamp_ms,
            "usage": {
                "provider": usage.provider,
                "model": usage.model,
                "resolvedRef": usage.resolved_ref,
                "api": usage.api,
                "transport": usage.transport,
                "inputTokens": usage.input_tokens,
                "outputTokens": usage.output_tokens,
                "cacheReadTokens": usage.cache_read_tokens,
                "cacheWriteTokens": usage.cache_write_tokens,
                "totalTokens": usage.total_tokens,
                "contextUsedTokens": usage.context_used_tokens,
                "contextTokenBudget": usage.context_token_budget,
                "durationMs": usage.duration_ms,
                "turnCost": usage.turn_cost,
                "currency": usage.currency,
            },
        }
        with self._lock:
            self._path.parent.mkdir(parents=True, exist_ok=True)
            if session_id in self._read_records():
                return False
            with self._path.open("a", encoding="utf-8", newline="\n") as handle:
                handle.write(json.dumps(record, ensure_ascii=False, separators=(",", ":")) + "\n")
        return True

    def totals(self, now_ms: int | None = None) -> UsageTotals:
        now = datetime.fromtimestamp((now_ms or round(datetime.now().timestamp() * 1000)) / 1000, self._timezone)
        totals = UsageTotals()
        if not self._path.exists():
            return totals
        currencies: set[Currency] = set()
        with self._lock:
            for record in self._read_records().values():
                stamp = datetime.fromtimestamp(record.timestamp / 1000, self._timezone)
                tokens = record.tokens
                cost = record.cost
                currency = record.currency
                totals.all_time_tokens += tokens
                totals.all_time_cost += cost
                if stamp.year == now.year and stamp.month == now.month:
                    totals.month_tokens += tokens
                    totals.month_cost += cost
                if stamp.date() == now.date():
                    totals.today_tokens += tokens
                    totals.today_cost += cost
                if currency is not None:
                    currencies.add(currency)
        totals.currency = next(iter(currencies)) if len(currencies) == 1 else None
        return totals

    def _read_records(self) -> dict[str, _LedgerRecord]:
        records: dict[str, _LedgerRecord] = {}
        if not self._path.exists():
            return records
        for line_number, raw in enumerate(
            self._path.read_text(encoding="utf-8").splitlines(),
            start=1,
        ):
            try:
                value = json.loads(raw)
                if not isinstance(value, dict):
                    continue
                runtime = value.get("runtime")
                record_id = value.get("id") or value.get("sessionId")
                timestamp = float(value["timestamp"])
                if runtime not in {"openclaw", "hermes"} or not isinstance(record_id, str):
                    continue
                usage_value = value.get("usage")
                usage = usage_value if isinstance(usage_value, dict) else {}
                tokens = _number(
                    value.get("tokens"),
                    usage.get("totalTokens"),
                    usage.get("total_tokens"),
                )
                if tokens is None:
                    tokens = (
                        (_number(usage.get("inputTokens"), usage.get("input_tokens")) or 0)
                        + (_number(usage.get("outputTokens"), usage.get("output_tokens")) or 0)
                        + (
                            _number(
                                usage.get("cacheReadTokens"),
                                usage.get("cache_read_tokens"),
                            )
                            or 0
                        )
                        + (
                            _number(
                                usage.get("cacheWriteTokens"),
                                usage.get("cache_write_tokens"),
                            )
                            or 0
                        )
                    )
                cost = (
                    _number(
                        value.get("cost"),
                        usage.get("turnCost"),
                        usage.get("turn_cost"),
                    )
                    or 0.0
                )
                raw_currency = value.get("currency") or usage.get("currency")
                currency: Currency | None = raw_currency if raw_currency in {"CNY", "USD"} else None
            except (KeyError, TypeError, ValueError, json.JSONDecodeError, OSError):
                continue
            key = record_id or f"legacy:{runtime}:{timestamp}:{line_number}"
            records[key] = _LedgerRecord(
                timestamp=timestamp,
                tokens=max(0, int(tokens)),
                cost=max(0.0, float(cost)),
                currency=currency,
            )
        return records


def _number(*values: object) -> float | None:
    for value in values:
        if isinstance(value, bool) or not isinstance(value, (int, float)):
            continue
        parsed = float(value)
        if isfinite(parsed):
            return parsed
    return None
